R/fixed_regex_linter.R

```diff
@@ -119,13 +101,82 @@ is_not_regex <- function(str) {
     ),
     "]"
   )
-  rx_static_token <- rex::rex(or(
+  rex::rex(or(
+    capture(rx_char_escape, name = "char_escape"),
+    capture(rx_trivial_char_group, name = "trivial_char_group")
+  ))
+})
+
+rx_static_token <- local({
+  rex::rex(or(
     rx_non_active_char,
-    rx_char_escape,
-    rx_trivial_char_group
+    rx_static_escape
   ))
-  rx_static_regex <- rex::rex(start, zero_or_more(rx_static_token), end)
+})
+
+rx_static_regex <- rex::rex(start, zero_or_more(rx_static_token), end)
+rx_first_static_token <- rex::rex(start, zero_or_more(rx_non_active_char), rx_static_escape)
 
+#' Determine whether a regex pattern actually uses regex patterns
+#'
+#' Note that is applies to the strings that are found on the XML parse tree,
+#'   _not_ plain strings. This is important for backslash escaping, which
+#'   happens at different layers of escaping than one might expect. So testing
+#'   this function is best done through testing the expected results of a lint
+#'   on a given file, rather than passing strings to this function, which can
+#'   be confusing.
+#'
+#' @param str A character vector.
+#' @return A logical vector, `TRUE` wherever `str` could be replaced by a
+#'   string with `fixed = TRUE`.
+#' @noRd
+is_not_regex <- function(str) {
   # need to add single-line option to allow literal newlines
-  grepl(paste0("(?s)", rx_static_regex), str, perl = TRUE) | grepl(rx_static_regex, str)
+  grepl(paste0("(?s)", rx_static_regex), str, perl = TRUE)
+}
+
+#' Compute a fixed string equivalent to a static regular expression
+#'
+#' @param static_regex A regex for which `is_not_regex()` returns `TRUE`
+#' @return A string such that `grepl(static_regex, x)` is equivalent to `grepl(get_fixed_string(static_regex), x, fixed = TRUE)`
+#'
+#' @noRd
+get_fixed_string <- function(static_regex) {
+  if (length(static_regex) == 0L) {
+    return(character())
+  } else if (length(static_regex) > 1L) {
+    return(vapply(static_regex, get_fixed_string, character(1L)))
+  }
+  fixed_string <- ""
```

Comment:
maybe more efficient to use list(nchar(static_regex)) (fixed version has to be weakly shorter right?) and do.call(paste0) at the end

Reply:
Not sure if that's efficient since we grossly over-allocate on most regexes.

Reply:
but it's just slot pointers overallocated. my instinct is overwriting is less efficient.

OTOH when I wrote this I hadn't read carefully enough -- thought we were building the fixed version char-by-char but it's actually chunk-wise, so I think it's fine